connectors/sources/microsoft_teams.py

```diff
@@ -0,0 +1,340 @@
+
+        self._token_cache_with_username.set_value(
+            access_token, now + timedelta(seconds=expires_in)
+        )
+
+        return access_token
+
+    @retryable(
+        retries=RETRY_COUNT,
+        interval=RETRY_INTERVAL,
+        strategy=RetryStrategy.EXPONENTIAL_BACKOFF,
+    )
+    async def _fetch_token(self):
+        """Generate API token for usage with Graph API
+
+        Returns:
+            (str, int) - a tuple containing access token as a string and number of seconds it will be valid for as an integer
+        """
+        authority = f"{GRAPH_API_AUTH_URL}/{self.tenant_id}"
+
+        auth_context = ConfidentialClientApplication(
+            client_id=self.client_id,
+            client_credential=self.client_secret,
+            authority=authority,
+        )
+        token_metadata = auth_context.acquire_token_by_username_password(
+            username=self.username, password=self.password, scopes=SCOPE
+        )
+        if not token_metadata.get("access_token"):
+            raise TokenFetchFailed(
+                f"Failed to authorize to Graph API. Please verify, that provided details are valid. Error: {token_metadata.get('error_description')}"
+            )
+
+        access_token = token_metadata.get("access_token")
+        expires_in = int(token_metadata.get("expires_in", TOKEN_EXPIRES))
+        return access_token, expires_in
+
+
+class MicrosoftTeamsClient:
+    """Client Class for API calls to Microsoft Teams"""
+
+    def __init__(self, tenant_id, client_id, client_secret, username, password):
+        self._sleeps = CancellableSleeps()
+        self._http_session = aiohttp.ClientSession(
+            headers={
+                "accept": "application/json",
+                "content-type": "application/json",
+            },
+            timeout=aiohttp.ClientTimeout(total=None),
+            raise_for_status=True,
+        )
+        self._api_token = GraphAPIToken(
+            tenant_id=tenant_id,
+            client_id=client_id,
+            client_secret=client_secret,
+            username=username,
+            password=password,
+        )
+
+        self._logger = logger
+
+    def set_logger(self, logger_):
+        self._logger = logger_
+
+    async def fetch(self, url):
+        return await self._get_json(absolute_url=url)
+
+    async def _get_json(self, absolute_url):
+        self._logger.debug(f"Fetching url: {absolute_url}")
+        async for response in self._get(absolute_url=absolute_url):
+            return await response.json()
+
+    @retryable(
+        retries=RETRY_COUNT,
+        interval=RETRY_INTERVAL,
+        strategy=RetryStrategy.EXPONENTIAL_BACKOFF,
+        skipped_exceptions=[NotFound, PermissionsMissing],
+    )
+    async def _get(self, absolute_url):
+        try:
+            token = await self._api_token.get_with_username_password()
+            self._logger.debug(f"Calling Microsoft Teams Endpoint: {absolute_url}")
```

Comment:
This is already logged in `_get_json`. I also feel there are too many levels of nesting: `fetch` -> `_get_json` -> `_get`

Reply:
Let me remove the logger `_get_json` logger as `_get` logger will be used in in indexing PR.
For having too many levels, I have kept it like this because in indexing PR we need to bifurcate `fetch` and `scroll` method separately on the basis of response as in some api we are getting response in `value` key and in some apis we are getting the response directly in a dictionary. Also for `get_json` we will use this in 2 methods `fetch` and `scroll`(part of indexing pr) that's why wrap the logic separately and using it .
To keep all the logic in single method we need to add flags and pass it all over the code. Let me know if you want to change it with the flags